src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/inlayhint/LSPInlayHintInlayProvider.java

```diff
@@ -66,47 +63,71 @@ public InlayHintsCollector getCollectorFor(@NotNull PsiFile psiFile,
+                    elements.forEach((offset, list) ->
+                            inlayHintsSink.addInlineElement(offset, false, toPresentation(editor, list, getFactory()), false));
                 } catch (InterruptedException e) {
                     LOGGER.warn(e.getLocalizedMessage(), e);
                     Thread.currentThread().interrupt();
                 }
                 return false;
             }
+
+            @NotNull
+            private List<Pair<Integer, Pair<InlayHint, LanguageServer>>> createInlayHints(
+                    @NotNull Document document,
+                    BlockingDeque<Pair<InlayHint, LanguageServer>> pairs,
+                    CompletableFuture<Void> future)
+                    throws InterruptedException {
+                List<Pair<Integer, Pair<InlayHint, LanguageServer>>> inlayHints = new ArrayList<>();
+                while (!future.isDone() || !pairs.isEmpty()) {
+                    ProgressManager.checkCanceled();
+                    Pair<InlayHint, LanguageServer> pair = pairs.poll(25, TimeUnit.MILLISECONDS);
+                    if (pair != null) {
+                        int offset = LSPIJUtils.toOffset(pair.getFirst().getPosition(), document);
+                        inlayHints.add(Pair.create(offset, pair));
+                    }
+                }
+                return inlayHints;
+            }
+
+            private CompletableFuture<Void> collect(@NotNull Project project, @NotNull Document document, InlayHintParams param, BlockingDeque<Pair<InlayHint, LanguageServer>> pairs) {
+                return LanguageServiceAccessor.getInstance(project)
+                        .getLanguageServers(document, capabilities -> capabilities.getInlayHintProvider() != null)
+                        .thenComposeAsync(languageServers -> CompletableFuture.allOf(languageServers.stream()
+                                .map(languageServer -> languageServer.getSecond().getTextDocumentService().inlayHint(param)
+                                        .thenAcceptAsync(inlayHints -> {
+                                            // textDocument/codeLens may return null
+                                            if (inlayHints != null) {
+                                                inlayHints.stream().filter(Objects::nonNull)
+                                                        .forEach(inlayHint -> pairs.add(new Pair(inlayHint, languageServer.getSecond())));
+                                            }
+                                        }))
+                                .toArray(CompletableFuture[]::new)));
+            }
         };
     }
 
-    private InlayPresentation toPresentation(Editor editor, int offset,
+    @NotNull
+    private static Range getViewPortRange(Editor editor) {
+        Position start = new Position(0, 0);
```

Comment:
which means you're likely unnecessarily computing hidden inlayhints

Reply:
I did that because the inlay hint is not refreshed when you are scrolling. Vscode provides this feature because when you scroll it refreshes the inlay hint. IJ doesn't support that it means that you need to consider that view port is the full document and not the visual view port.

Reply:
please add a comment then, as it's not obvious

Reply:
OK I have add a comment